### PR TITLE
fix(http): send error to http client instead of closing connection

### PR DESCRIPTION
**Pull request by glasstiger**
Currently if an error occurs after the HTTP has already been written into the HTTP response, the server simply cuts the connection.
This usually results in the client receiving HTTP 200 in the header and likely in an empty response body.
The Web Console, for example, does not realise at all that there was a problem on the server side.
It simply shows an empty grid, although the query failed.
This change fixes the above problem by adding the error to the query suffix.
The Web Console is already parsing this field from the response json, and interprets it as an error.

required by https://github.com/questdb/ui/pull/348

### CODE REVIEW

**Comment by questdb-butler** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Please update the PR title to match this format:
`type(subType): description`

Where `type` is one of:
`feat`, `fix`, `chore`, `docs`, `style`, `refactor`, `perf`, `test`, `ci`, `revert`

And: `subType` is one of:
`build`, `sql`, `log`, `mig`, `core`, `ilp`, `pgwire`, `http`, `conf`, `ui`, `wal`

For Example:

```
perf(sql): improve pattern matching performance for SELECT sub-queries
```
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against f320ac30ebcff8b6839806cb76c7b437225dbe83
</p>

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 96 / 103 (93.20%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessor.java|13|16|81.25%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/SimulateCrashFunctionFactory.java|13|15|86.67%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessorState.java|44|46|95.65%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpResponseSink.java|25|25|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|

**Review by ideoma — COMMENTED** [review]
It works, OK, I can simulate the error adding NPE into long_sequence cursor. However when error is sent back I see 
`HTTP 500 (Internal server error)` in the web console
despite having HTTP response code 200 

image

image

**Comment by glasstiger** [comment]
> It works, OK, I can simulate the error adding NPE into long_sequence cursor. However when error is sent back I see 
> 
> `HTTP 500 (Internal server error)` in the web console
> 
> despite having HTTP response code 200 
> 
> 
> 
> image
> 
> 
> 
> image
> 
>  HTTP response code should be the same as in the error code, 500 in this case
> 
> 

This is expected.
The 200 cannot be changed, because the error happens after the header  has been written to the socket.

That is why we previously disconnected the client.
However, that left the web console think that it is a simple empty response.
Now it knows that although the header has 200, there is a problem.

**Comment by glasstiger** [comment]
> It works, OK, I can simulate the error adding NPE into long_sequence cursor. However when error is sent back I see
> `HTTP 500 (Internal server error)` in the web console
> despite having HTTP response code 200
> image
> image
> HTTP response code should be the same as in the error code, 500 in this case

What if I did not add the `HTTP 500` part to the error message?
That could clear up maybe the confusion.
Basically the error message would be just `Internal server error`, or it could be just the message from the exception, although a NPE message, for example, will not look nice.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 98 / 105 (93.33%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessor.java|13|16|81.25%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/SimulateCrashFunctionFactory.java|15|17|88.24%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessorState.java|44|46|95.65%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpResponseSink.java|25|25|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 99 / 105 (94.29%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessor.java|14|16|87.50%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/SimulateCrashFunctionFactory.java|15|17|88.24%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessorState.java|44|46|95.65%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpResponseSink.java|25|25|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|

**Review by bluestreak01 — DISMISSED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 100 / 105 (95.24%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessor.java|13|16|81.25%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/SimulateCrashFunctionFactory.java|15|17|88.24%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpResponseSink.java|25|25|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessorState.java|46|46|100.00%|

**Comment by bluestreak01** [comment]
@glasstiger new HTTP tests should be written using httpClient, check `assertGet()`

**Review by bluestreak01 — CHANGES_REQUESTED** [review]
buffer overflow still causes disconnect

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 100 / 116 (86.21%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessor.java|20|27|74.07%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessorState.java|44|51|86.27%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/SimulateCrashFunctionFactory.java|15|17|88.24%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpResponseSink.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 102 / 116 (87.93%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessor.java|20|27|74.07%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/SimulateCrashFunctionFactory.java|15|17|88.24%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessorState.java|46|51|90.20%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpResponseSink.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 107 / 124 (86.29%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessor.java|20|27|74.07%|
|:large_blue_circle:|io/questdb/cairo/CairoException.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessorState.java|44|51|86.27%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/SimulateCrashFunctionFactory.java|15|17|88.24%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpConnectionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpResponseSink.java|21|21|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|2|2|100.00%|

**Comment by glasstiger** [comment]
> buffer overflow still causes disconnect

fixed

**Comment by glasstiger** [comment]
> @glasstiger new HTTP tests should be written using httpClient, check `assertGet()`

fixed

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 109 / 124 (87.90%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessor.java|20|27|74.07%|
|:large_blue_circle:|io/questdb/cairo/CairoException.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/SimulateCrashFunctionFactory.java|15|17|88.24%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessorState.java|46|51|90.20%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpConnectionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpResponseSink.java|21|21|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|2|2|100.00%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 107 / 124 (86.29%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessor.java|20|27|74.07%|
|:large_blue_circle:|io/questdb/cairo/CairoException.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessorState.java|44|51|86.27%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/SimulateCrashFunctionFactory.java|15|17|88.24%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpConnectionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpResponseSink.java|21|21|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|2|2|100.00%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 107 / 124 (86.29%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessor.java|20|27|74.07%|
|:large_blue_circle:|io/questdb/cairo/CairoException.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessorState.java|44|51|86.27%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/SimulateCrashFunctionFactory.java|15|17|88.24%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpConnectionContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpResponseSink.java|21|21|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|2|2|100.00%|

File: core/src/main/java/io/questdb/cairo/CairoException.java [hunk 1]
```diff
@@ -331,7 +324,6 @@ protected void clear(int errno) {
         cacheable = false;
         interruption = false;
         authorizationError = false;
-        entityDisabled = false;
```
Comment (amunra) [inline core/src/main/java/io/questdb/cairo/CairoException.java:334]: I think I'm lacking context here, but why is the field removed?

Comment (glasstiger) [reply core/src/main/java/io/questdb/cairo/CairoException.java:334]: > I think I'm lacking context here, but why is the field removed?

Not needed anymore.
The piece code, which used `CairoException::isEntityDisabled()` has been removed.

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**